crypto/cmp/cmp_protect.c

```diff
@@ -0,0 +1,315 @@
+                || (protection = OPENSSL_malloc(sig_len)) == NULL
+                || EVP_DigestSignFinal(evp_ctx, protection, &sig_len) <= 0) {
+            CMPerr(0, CMP_R_ERROR_CALCULATING_PROTECTION);
+            goto end;
+        }
+    } else {
+        CMPerr(0, CMP_R_INVALID_ARGS);
+        goto end;
+    }
+
+    if ((prot = ASN1_BIT_STRING_new()) == NULL)
+        goto end;
+    /* OpenSSL defaults all bit strings to be encoded as ASN.1 NamedBitList */
+    prot->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07);
+    prot->flags |= ASN1_STRING_FLAG_BITS_LEFT;
+    if (!ASN1_BIT_STRING_set(prot, protection, sig_len)) {
+        ASN1_BIT_STRING_free(prot);
+        prot = NULL;
+    }
+
+ end:
+    OSSL_CRMF_PBMPARAMETER_free(pbm);
+    EVP_MD_CTX_free(evp_ctx);
+    OPENSSL_free(protection);
+    OPENSSL_free(prot_part_der);
+    return prot;
+}
+
+int ossl_cmp_msg_add_extraCerts(OSSL_CMP_CTX *ctx, OSSL_CMP_MSG *msg)
+{
+    if (!ossl_assert(ctx != NULL && msg != NULL))
+        return 0;
+
+    if (msg->extraCerts == NULL
+            && (msg->extraCerts = sk_X509_new_null()) == NULL)
+        return 0;
+
+    if (ctx->clCert != NULL) {
+        /* Make sure that our own cert gets sent, in the first position */
+        if (!X509_up_ref(ctx->clCert))
+            return 0;
+        if (!sk_X509_push(msg->extraCerts, ctx->clCert)) {
+            X509_free(ctx->clCert);
+            return 0;
+        }
+        /* if we have untrusted store, try to add intermediate certs */
+        if (ctx->untrusted_certs != NULL) {
+            STACK_OF(X509) *chain =
+                ossl_cmp_build_cert_chain(ctx->untrusted_certs, ctx->clCert);
+            int res = ossl_cmp_sk_X509_add1_certs(msg->extraCerts, chain,
+                                                  1 /* no self-signed */,
+                                                  1 /* no duplicates */, 0);
+            sk_X509_pop_free(chain, X509_free);
+            if (res == 0)
+                return 0;
+        }
+    }
+
+    /* add any additional certificates from ctx->extraCertsOut */
+    if (!ossl_cmp_sk_X509_add1_certs(msg->extraCerts, ctx->extraCertsOut, 0,
+                                     1 /* no duplicates */, 0))
+        return 0;
+
+    /* if none was found avoid empty ASN.1 sequence */
+    if (sk_X509_num(msg->extraCerts) == 0) {
+        sk_X509_free(msg->extraCerts);
+        msg->extraCerts = NULL;
+    }
+    return 1;
+}
+
+/*
+ * Create an X509_ALGOR structure for PasswordBasedMAC protection based on
+ * the pbm settings in the context
+ * returns pointer to X509_ALGOR on success, NULL on error
+ */
+static X509_ALGOR *create_pbmac_algor(OSSL_CMP_CTX *ctx)
+{
+    X509_ALGOR *alg = NULL;
+    OSSL_CRMF_PBMPARAMETER *pbm = NULL;
+    unsigned char *pbm_der = NULL;
+    int pbm_der_len;
+    ASN1_STRING *pbm_str = NULL;
+
+    if (!ossl_assert(ctx != NULL))
+        return NULL;
+
+    if ((alg = X509_ALGOR_new()) == NULL)
```

Comment:
I do personally not like the style with
```C
if ((bla = bla_new()) == NULL)
  goto err;
if ((blub = blub_new()) == NULL)
  goto err;
```
I do not see what is the advantage to
```C
bla = bla_new();
blub = blub_nue();
if (bla == NULL || blub == NULL)
  goto err;
```

Reply:
In this case you are right, it would work in the proposed way. But in some other cases an allocated object is used to initialize the next object. So generally the "fail-fast" coding style is better for robust error handling and  error tracking.

Reply:
I do not agree that this is better style in every case, I would like the
alternatives considered at least, and the best alternative used when
appropriate. As an example I tried this:
```C
#include <stdio.h>
int main()
{
  int rv=0;
  if (!printf("hallo "))
    rv = 1;
  if (!printf("WELT\n"))
    rv = 1;
  if (!printf("hallo ") 
      || !printf("WELT\n"))
    rv = 1;
 return rv;
}
```
complie at -O2 and try to debug this:
```
$ gcc -g -O2 tt.c
$ gdb ./a.out
(gdb) b main
Haltepunkt 1 at 0x401060: file tt.c, line 5.
(gdb) r
Starting program: /home/ed/gnu/gcc-test/a.out 

Breakpoint 1, main () at tt.c:5
5	  if (!printf("hallo "))
(gdb) n
7	  if (!printf("WELT\n"))
(gdb) n
hallo WELT
9	  if (!printf("hallo ") 
(gdb) n
hallo WELT
0x00007ffff7a32f45 in __libc_start_main () from /lib/x86_64-linux-gnu/libc.so.6
(gdb) 
```
you see the logic in the if statement is not extra debug-friendly,
especially when it is more complicated than necessary.
While simple assignments tend to be easier to debug.

Reply:
I changed it.